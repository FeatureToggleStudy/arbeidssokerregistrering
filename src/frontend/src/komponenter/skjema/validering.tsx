import * as React from 'react';
import { rules } from 'react-redux-form-validation';
import { FormattedMessage } from 'react-intl';

export function paakrevdTekst(feilmeldingId: string) {
    return rules.minLength(0, <FormattedMessage id={feilmeldingId} />);
}

interface FeedbackSummaryCreatorProps {
    header: React.ReactNode;
    errors: React.ReactNode[];
}

export function FeedbackSummaryCreator({ header, errors }: FeedbackSummaryCreatorProps) {
    return (
        <div
            role="alert"
            aria-live="assertive"
            aria-atomic="true"
            className="skjema__feilomrade--harFeil"
            tabIndex={-1}
        >
            <h3>{header}</h3>
            <ul>
                {errors}
            </ul>
        </div>
    );
}