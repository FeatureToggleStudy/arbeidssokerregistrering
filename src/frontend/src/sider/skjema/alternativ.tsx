import * as React from 'react';
import { Radio } from 'nav-frontend-skjema';
import InjectedIntlProps = ReactIntl.InjectedIntlProps;

interface AlternativProps {
    hentAvgittSvar: () => number | undefined;
    alternativId: number;
    avgiSvar: (alternativId: number) => void;
    getTekstId: (alternativId: number) => string;
}

function Alternativ(props: AlternativProps & InjectedIntlProps) {
    const tekst = props.intl.messages[props.getTekstId(props.alternativId)];
    return (
        <Radio
            onChange={() => props.avgiSvar(props.alternativId)}
            className="blokk-xs"
            name={'alternativ'}
            label={tekst}
            value={tekst}
            checked={props.hentAvgittSvar() === props.alternativId}
        />
    );
}

export default Alternativ;