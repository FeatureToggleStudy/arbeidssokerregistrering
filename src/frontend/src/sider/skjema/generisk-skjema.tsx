import * as React from 'react';
import KnappNeste from '../../komponenter/knapper/knapp-neste';
import Knapperad from '../../komponenter/knapper/knapperad';
import KnappAvbryt from '../../komponenter/knapper/knapp-avbryt';
import Tilbakeknapp from '../../komponenter/knapper/tilbakeknapp';
import {erSelvgaende} from "./sporsmal-utils";

interface GeneriskSkjemaProps {
    children: {};
    gjeldendeSporsmal: number;
    sporsmalErBesvart: (sporsmalId: string) => boolean;
    hentAvgittSvar: (sporsmalId: string) => number | undefined;
    avbrytSkjema: () => void;
    gaaTilSporsmal: (sporsmal: number) => void;
    gaaTilbake: () => void;
    fullforSkjema: () => void;
    gaaTilSblRegistrering: () => void;
}

type Props = GeneriskSkjemaProps;

export default class GeneriskSkjema extends React.Component<Props> {

    private antallSporsmal: number;
    private sporsmalIder: string[];

    constructor(props: Props) {
        super(props);
    }

    render() {
        this.antallSporsmal = React.Children.toArray(this.props.children).length;
        const gjeldendeSporsmalComponent = this.props.children[this.props.gjeldendeSporsmal - 1];
        this.sporsmalIder = this.getSporsmalIder();
        const disableKnappNeste = (this.props.gjeldendeSporsmal === this.antallSporsmal) ?
            !this.alleSporsmalErBesvarte() :
            !this.props.sporsmalErBesvart(this.sporsmalIder[this.props.gjeldendeSporsmal]);

        return (
            <React.Fragment>
                <div className="blokk panel-skjema-wrapper" >
                    <Tilbakeknapp onClick={() => this.props.gaaTilbake()}/>
                    {gjeldendeSporsmalComponent}
                </div>

                <Knapperad>
                    <KnappAvbryt
                        classname="knapp mmr"
                        onClick={() => this.props.avbrytSkjema()}
                    />
                    <KnappNeste
                        onClick={() => this.gaaTilNesteSporsmal()}
                        disabled={disableKnappNeste}
                    />
                </Knapperad>
            </React.Fragment>
        );
    }

    gaaTilNesteSporsmal() {
        const gjeldendeSporsmalId = this.sporsmalIder[this.props.gjeldendeSporsmal];
        if (erSelvgaende(gjeldendeSporsmalId, this.props.hentAvgittSvar(gjeldendeSporsmalId))) {
            this.props.gaaTilSblRegistrering();
        }
        if (this.sporsmalIder[this.props.gjeldendeSporsmal])
        if ((this.props.gjeldendeSporsmal === this.antallSporsmal) && this.alleSporsmalErBesvarte()) {
            this.props.fullforSkjema();
        } else {
            this.props.gaaTilSporsmal(this.props.gjeldendeSporsmal + 1);
        }
    }

    alleSporsmalErBesvarte(): boolean {
        const besvarteSporsmal = this.sporsmalIder.filter(sporsmalId => this.props.sporsmalErBesvart(sporsmalId));
        return besvarteSporsmal.length === this.antallSporsmal;
    }

    getSporsmalIder(): string[] {
        let map: string[] = ['']; // Det første spørsmålet skal korrespondere med 1, ikke 0
        for (let i = 0; i < this.antallSporsmal; i += 1) {
            map.push(this.props.children[i].props.sporsmalId);
        }
        return map;
    }
}