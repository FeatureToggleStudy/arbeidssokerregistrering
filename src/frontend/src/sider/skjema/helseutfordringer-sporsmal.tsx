
interface Props {
}

export default function HelseutfordringerSporsmal({}: Props) {
    return (null);
}

/*



<Skjema>
    <Helsesporsmal>
        <HelseutfordringerSporsmal navn="helse"/>
        <UtdanningSporsmal navn="utdanning"/>
    </Helsesporsmal>
    <nesteknapp/>
</Skjema>


<Helsesporsmal>
    <Systemtittel> tittel </Systemtittel>
    <Alternativ> alternativ1 </Alternativ>
    <Alternativ> alternativ1 </Alternativ className="kult-spm">
</Helsesporsmal>


state.svar = [
{
    sporsmal: "helse"
    svar: 2
},
{
    sporsmal: "utdanning"
    svar: 4
}
]

 */