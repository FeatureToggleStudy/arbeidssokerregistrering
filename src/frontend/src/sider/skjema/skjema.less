.knapp-neste {
    right:0
}

.spm-tittel {
    padding-top: @padding-small-horizontal;
    padding-bottom: @padding-small-horizontal;
    text-align: center;
}
.sporsmal-wrapper {
    margin-left: auto;
    margin-right: auto;
    max-width: @screen-sm-min;
    &:focus {
        outline: 0;
    }
}

.skjema-container-wrapper:focus {
    outline: 0;
}

.panel-skjema {
    padding-bottom: 0.1rem;
}

/*
.form-skjema {
    max-width: 50%;
    margin: 0 auto;
}
*/
.form-skjema {
    max-width: 50rem;
    margin-left: auto;
    margin-right: auto;
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
}

.form-skjema .alternativ-wrapper {
    width: 20rem;
}

.alternativ-wrapper .radioPanel {
    width: 100%;
}

.alternativ-wrapper {
    margin: 0.25rem;
}

@media (max-width: @screen-sm-min) {
    .sporsmal-wrapper {
        margin-left: 2rem;
        margin-right: 2rem;
    }
    .form-skjema {
        max-width: inherit;
    }
    .form-flex {
        display: flex;
        justify-items: flex-start;

        .skjemaelement {
            margin-right: 2rem;
        }
    }
}

