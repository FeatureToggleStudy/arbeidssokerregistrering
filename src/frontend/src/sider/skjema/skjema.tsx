import * as React from 'react';
import KnappNeste from '../../komponenter/knapper/knapp-neste';
import Knapperad from '../../komponenter/knapper/knapperad';
import KnappAvbryt from '../../komponenter/knapper/knapp-avbryt';
import Tilbakeknapp from '../../komponenter/knapper/tilbakeknapp';

interface SkjemaProps {
    children: {};
    gjeldendeSporsmal: number;
    sporsmalErBesvart: (sporsmalId: string) => boolean;
    avbrytSkjema: () => void;
    gaaTilbake: () => void;
    gaaTilNesteSide: (gjeldendeSporsmalId: string, alleSporsmalIder: string[]) => void;
}

type Props = SkjemaProps;

export default class Skjema extends React.Component<Props> {

    private antallSporsmal: number;
    private sporsmalIder: string[];

    constructor(props: Props) {
        super(props);
    }

    render() {
        this.antallSporsmal = React.Children.toArray(this.props.children).length;
        const gjeldendeSporsmalComponent = this.props.children[this.props.gjeldendeSporsmal];
        this.sporsmalIder = this.getSporsmalIder();
        const disableKnappNeste = (this.props.gjeldendeSporsmal === this.antallSporsmal) ?
            !this.alleSporsmalErBesvarte() :
            !this.props.sporsmalErBesvart(this.sporsmalIder[this.props.gjeldendeSporsmal]);

        return (
            <React.Fragment>
                <Tilbakeknapp onClick={() => this.props.gaaTilbake()}/>
                {gjeldendeSporsmalComponent}

                <Knapperad>
                    <KnappAvbryt
                        classname="knapp mmr"
                        onClick={() => this.props.avbrytSkjema()}
                    />
                    <KnappNeste
                        onClick={() => this.nesteButtonClick()}
                        disabled={disableKnappNeste}
                    />
                </Knapperad>
            </React.Fragment>
        );
    }

    nesteButtonClick() {
        const gjeldendeSporsmalId = this.sporsmalIder[this.props.gjeldendeSporsmal];
        this.props.gaaTilNesteSide(gjeldendeSporsmalId, this.getSporsmalIder());
    }

    alleSporsmalErBesvarte(): boolean {
        const besvarteSporsmal = this.sporsmalIder.filter(sporsmalId => this.props.sporsmalErBesvart(sporsmalId));
        return besvarteSporsmal.length === this.antallSporsmal;
    }

    getSporsmalIder(): string[] {
        let map: string[] = [];
        for (let i = 0; i < this.antallSporsmal; i += 1) {
            map.push(this.props.children[i].props.sporsmalId);
            // For at this.props.children[i] skal funke trenger man minst to children
        }
        return map;
    }
}