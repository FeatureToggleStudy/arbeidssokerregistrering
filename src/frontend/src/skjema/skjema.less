.knapp-neste {
    right:0
}