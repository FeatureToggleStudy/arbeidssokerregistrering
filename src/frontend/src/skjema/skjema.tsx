import * as React from 'react';
import { connect, Dispatch } from 'react-redux';
import { InjectedIntlProps, injectIntl } from 'react-intl';
import antallSporsmal from '../sporsmal/alle-sporsmal';
import { Undertittel } from 'nav-frontend-typografi';
import { Panel } from 'nav-frontend-paneler';
import { endreSvarAction } from '../ducks/svar';
import { AppState } from '../reducer';
import Alternativ, { EndreSvar } from './alternativ';
import { RouteComponentProps } from 'react-router';
import KnappNeste from './knapp-neste';
import KnappFullfor from './knapp-fullfor';
import { configSpmPrSide } from './skjema-utils';

interface SkjemaProps {
    id: string;
    resetSvar: () => void;
}

export interface MatchProps {
    id: string;
}

interface StateProps {
    sporsmalErBesvart: (sporsmalId: string) => boolean;
    hentAvgittSvarId: (sporsmalId: string) => string;
}

interface DispatchProps {
    endreSvar: EndreSvar;
}

type Props = SkjemaProps & InjectedIntlProps & DispatchProps & StateProps & RouteComponentProps<MatchProps>;

function Skjema({match, history, intl, endreSvar, sporsmalErBesvart, hentAvgittSvarId}: Props) {
    const sideId = match.params.id;
    // const antallAlternativer = antallSporsmal[parseInt(sporsmalId, 10) - 1];
    // const tittelId = `sporsmal-${sporsmalId}-tittel`;
    // const avgittSvarId = parseInt(hentAvgittSvarId(sideId), 10);

    // todo
    // test send tilbake til fÃ¸rsteside, dersom sideurl ikke finnes
    // test fullforknap disable frem til alle spm er besvart

    const spmListePaSiden = configSpmPrSide[sideId];

    let disbl = true;
    spmListePaSiden.map((spmId: string) => {
        disbl = !sporsmalErBesvart(spmId);
    });

    return (
        <div>
            {
                spmListePaSiden
                    .map((spmId: string) => (
                        <div>
                            <Undertittel className="blokk-xxs">
                                {intl.messages[`sporsmal-${spmId}-tittel`]}
                            </Undertittel>
                            <Panel className="blokk-s">
                                <form >
                                    {Array.from(Array(antallSporsmal[parseInt(spmId, 10) - 1]).keys())
                                        .map(i => i + 1)
                                        .map((key) => <Alternativ
                                            sporsmalId={spmId}
                                            endreSvar={endreSvar}
                                            key={key}
                                            alternativId={key.toString()}
                                            tekstId={`sporsmal-${sideId}-alternativ-${key}`}
                                            checked={key === parseInt(hentAvgittSvarId(spmId), 10)}
                                            intl={intl}
                                        />)}
                                </form>
                            </Panel>
                        </div>
                    ))
            }

            <div className="skjema-knapper">
                {
                    parseInt(sideId, 10) === Object.keys(configSpmPrSide).length ?
                        <KnappFullfor
                            disabled={disbl}
                            onClick={() => history.push('/oppsummering')}
                        />
                        :
                        <KnappNeste
                            disabled={disbl}
                            onClick={(() => {
                                history.push(`/skjema/${(parseInt(sideId, 10) + 1)}`);
                            })}
                        />
                }
            </div>
        </div>
    );
}

const mapStateToProps = (state: AppState): StateProps => ({
    sporsmalErBesvart: (sporsmalId) => !!state.svar[sporsmalId],
    hentAvgittSvarId: (sporsmalId) => state.svar[sporsmalId]
});

const mapDispatchToProps = (dispatch: Dispatch<AppState>): DispatchProps => ({
    endreSvar: (sporsmalId, alternativId) => dispatch(endreSvarAction(sporsmalId, alternativId)),
});

export default connect(mapStateToProps, mapDispatchToProps)
(injectIntl(Skjema)
);