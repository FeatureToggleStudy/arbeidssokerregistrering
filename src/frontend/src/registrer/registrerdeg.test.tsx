import * as React from 'react';
import IntlProvider from '../Intl-provider';
import * as renderer from 'react-test-renderer';
import Registrerdeg from './registrerdeg';
import { MatchProps } from '../skjema/skjema';
import { RouteComponentProps } from 'react-router';

test('Registrerdeg komponent har riktig innhold og struktur', () => {
    const component = renderer.create(
        <IntlProvider><Registrerdeg {...{} as RouteComponentProps<MatchProps>}  /></IntlProvider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});