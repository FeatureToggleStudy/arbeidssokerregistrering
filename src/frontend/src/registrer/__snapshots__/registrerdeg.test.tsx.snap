// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registrerdeg komponent har riktig innhold og struktur 1`] = `
<div>
  <div
    className="panel panel-info blokk-l"
  >
    <h1
      className="typo-sidetittel overskrift-panel-info info-sirkel-bla"
    >
      <span>
        Registrer deg som arbeidssøker
      </span>
    </h1>
    <p
      className="typo-normal blokk-xs"
    >
      <span>
        Når du registrerer deg får flere fordeler:
      </span>
    </p>
    <ul
      className="typo-normal blokk-xs"
    >
      <li>
        <span>
          du får hjelp til å komme i jobb
        </span>
      </li>
      <li>
        <span>
          du kan søke om dagpenger eller annen økonomisk støtte
        </span>
      </li>
      <li>
        <span>
          arbeidsgivere finner deg i CV-basen vår
        </span>
      </li>
      <li>
        <span>
          du kan abonnere på ledige stillinger
        </span>
      </li>
    </ul>
    <p
      className="typo-normal registrerdeg-tips blokk-xl"
    >
      <span>
        Tips: Fullfør registrering først, så kan du søke om dagpenger.
      </span>
    </p>
  </div>
  <div
    className="panel-info__knapperad"
  >
    <button
      className="knapp knapp knapp--hoved knapp--hoved"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      <span>
        Start registrering
      </span>
    </button>
  </div>
</div>
`;
